### ISSUE TITLE
getStaticPaths with fallback true: Allow custom HTTP status codes

### ISSUE THREAD

**Issue by claus** [vercel/next.js#11646]
# Feature request

## Is your feature request related to a problem? Please describe.

Using getStaticPaths with `fallback: true`, pages that are not SSG'ed yet get served with HTTP status 200 no matter what, which makes sense as the response is immediate and getStaticProps is then called in the background.

However, i need a way to:
1. server side generate static pages on demand (e.g. i have a lot of product pages and can't generate them all at build time), AND:
2. serve a proper 404 status code for product page urls that don't exist

## Describe the solution you'd like

I should be able to configure the "fallback" mode so that it's possible that the response is _not_ immediate and Next instead waits for getStaticProps to return and only _then_ sends the response, allowing me to set the HTTP status code.

For example, instead of only having one `fallback` prop, have an `onDemand` prop and a `fallback prop`.
- `onDemand`: If true enables on-demand generation of static pages. If false, all pages returned by `getStaticPaths` are generated at build time and any other pages return 404s.
- `fallback`: If true enables fallback mode as it is currently implemented. If false, for pages that are not generated yet a response is sent only after the Promise returned by getServerProps resolved, allowing tweaks to a server response (e.g. setting `statusCode`). Only has any effect if `onDemand` is true.

Of course that would not be backwards compatible so it needs some thought. But i guess something like this could work, no?

## Describe alternatives you've considered

n/a

## Additional context

#11644

**Comment by ivan-kleshnin** [comment]
I support your proposal but not the solution. It's way overcomplicated for such a small task IMO. 

I think it'd be better to prodive some proxy/fake `res` object as `getStaticProps` argument with an option to set 404 and possibly some HTTP headers.

**Comment by jon64digital** [comment]
I have just run into the same problem and I believe this thread answers my third question here...

https://github.com/vercel/next.js/discussions/13715

The way I see it is that the server should never return a 200 code before it knows the page is actually 301 or 404.

The solution will never be perfect but I'd suggest that when using getStaticProps and fallback:true then any page which isn't previously known about must act like a server rendered page the first time  it's hit, or when the cache is stale. There would still be a SSG version of the page cached after the first hit. That's not unreasonable is it?

I mean, you still have the option of SSG at build time and getStaticPaths if you can use it, this is only for pages which aren't known about in advance which to be honest currently still have to be generated behind a loading indicator. So I guess the solution is "allow us to switch off the loading indicator and wait for a proper server response the first time we generate an unknown page"?

**Comment by jon64digital** [comment]
Still thinking about this one and it's the one issue stopping me from using SSG for most of my pages with CMS content at the moment.

My proposed solution is to have two different options which are mutually exclusive;

**fallback:true** 

Behaves as it does now. If a page has already been generated it is served, otherwise we fall back to a loading screen and the browser gets a 200 status code. In my opinion this is broken and nobody should use it but it has to stay for backwards compatibility.

**fallback server:true**

If the page has been previously generated then we're served the static file. If not then getStaticProps behaves exactly like getServerProps and "falls back" on a server load rather than a JavaScript loading indicator. Correct 404s and 301/302s could be served.

The second option would still serve a stale static version and regenerate in the background if the "revalidate" time has passed so the SSR version will only happen the first time somebody visits a page.

Personally, I'd only ever use option 2 but I guess some people might still use option 1 if they don't care about SEO.

**Comment by c58** [comment]
I'm building a SEO oriented CMS for my organization and option 2 from the previous comment is exactly what i need.

**Comment by BernardA** [comment]
@jon64digital, @c58 If I understand correctly what you are saying, that proposal would not solve this issue.

In my use case, I have a small ads app with about 200 categories that may potentially exist in about 20,000 cities. So it's not practical to pre-build all.

So, thinking about SEO, my URLs would look like:

`/search` for the main search page at `/pages/search/index.js` with data fetched either with `getStaticProps` and then refreshed in `componentDidMount` of directly with `getServerSideProps`, `stale while revalidate` strategy. 

`/search/[city]/[rootCategory]/[parentCategory]/[category] at `pages/search/[rootCategory]/[parentCategory/[category].js`
with data fetched with `getStaticPaths` and `getStaticProps`.

For the second case above, I am thinking about pre-building a subset of the most likely city/category combination. 
Those would have data fetched at build time and served `stale while revalidate` strategy, ie, refreshed at `componentDidMount`. I have not tested that yet, but in principle this should work. 

Then for the rest:

1 - user has entered valid information for all parameters => means that a **valid response** should be provided either with the available ads or with UI informing user that no ads currently exist. 
**So, having no ads is a perfectly valid response**. 
   - show initially the fallback UI while data is being fetched and then either show the ads or UI informing that there's no ads

2- use has entered invalid information for any of the parameters => should be redirected to 404.
I have found no way to redirect to 404 before `componentDidMount` which is certainly not ideal. 

The other concern I have in this case is that [according to the documentation](https://nextjs.org/docs/basic-features/data-fetching), `At the same time, Next.js adds this path to the list of pre-rendered pages`.

Of course that cannot happen for case 2 above, otherwise one would just be creating SEO issues.

**Comment by ivan-kleshnin** [comment]
https://nextjs.org/blog/next-9-4 claims as a benefit:

> Pages never go offline. If the background page re-generation fails, the old page remains unaltered.

But what if I want to remove some page... Is it impossible? 😨

**Comment by Timer** [comment]
We've just shipped experimental support for Blocking ("SSR") Fallback (in lieu of the static placeholder fallback)!

See https://github.com/vercel/next.js/pull/15672 for usage.

This does not yet cover the 404 use case, which is coming 🔜.

**Comment by claus** [comment]
@Timer I saw it. That's amazing, thank you so much!

**Comment by javiercr** [comment]
@Timer awesome 👏 ! Looking forward for the 404 support!

**Comment by jamesknelson** [comment]
> This does not yet cover the 404 use case, which is coming 🔜.

@Timer It'd be really helpful if it supports 301 redirects as well as 404s. E.g. for things like blog posts where the slug has been renamed after publication.

**Comment by timneutkens** [comment]
> @Timer It'd be really helpful if it supports 301 redirects as well as 404s. E.g. for things like blog posts where the slug has been renamed after publication.

See #14890

**Comment by lachlanjc** [comment]
There's support for `notFound` as well as `redirect` now: https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

**Comment by timneutkens** [comment]
Yep this has been solved, thanks @lachlanjc 👍

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
